Feature/request scoped event emitters

thiagomini opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

Request Scoped listeners are not available right now.


## What is the new behavior?

We can inject the `EVENT_PAYLOAD` symbol and make our listeners Request Scoped:

```ts
class EventListener {
  constructor(@Inject(EVENT_PAYLOAD) private readonly eventPayload: any) { }
}
```


## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No


## Other information

You can follow the commits to understand how it became from the first version to the refactored one, which I used to reduce the complexity of the `EventSubscribersLoader` class.

thiagomini commented on tests/src/events-provider.request-scoped.consumer.ts:
```diff
@@ -0,0 +1,24 @@
+    }
+  }
+
+  public static injectedEventPayload = {};
```
The reason why I created this static property is to make it easy to check the event payload was injected in the tests. Otherwise we would need a specific `contextId`

thiagomini commented on tsconfig.json:
```diff
@@ -9,7 +9,7 @@
     "experimentalDecorators": true,
     "target": "es6",
-    "sourceMap": false,
+    "sourceMap": true,
```
Changed this to enable local debug. It was really helpful in my experience here

delucca commented on lib/event-subscribers.loader.ts:
```diff
@@ -31,39 +45,126 @@ export class EventSubscribersLoader
+              instance,
+              methodKey,
+              isRequestScoped,
+              wrapper.host as Module,
```
There is no other way of doing this without using `as`? Just to avoid losing type-safety

thiagomini replied:
I tried a couple of things but apparently this is the less hacky way of doing it. To give you more context on that, the reason Why I had to do it is because of the following:

1. The `InstanceWrapper.host` is an optional attribute, as you can see [here](https://github.com/nestjs/nest/blob/c2d73f91d7848994302fb739f61c080e9e7f9b20/packages/core/injector/instance-wrapper.ts#L59)

2. This `host` attribute is used by the `injector.loadPerContext` method [in this same file](https://github.com/thiagomini/event-emitter/blob/a2bbb696968b5d95077d8f4a469a2bb8c9ed974f/lib/event-subscribers.loader.ts#L129)

3. `loadPerContext` method requires a `module` attribute, hence it's not optional

4. this project has strict flag set to true, as you can see in the [tsconfig.json](https://github.com/thiagomini/event-emitter/blob/26426c20abc08cf05cb98cb397f625489def805d/tsconfig.json#L5)

The reason **4** is the actual culprit of this `as Module` statement. You can see in the [nestjs/bull package](https://github.com/nestjs/bull/blob/87aba5742d40853d50b4a10dc4d63f4e7352ee39/packages/bull/lib/bull.explorer.ts#L81) the module is handed over the function with no casting, and that's because its [tsconfig](https://github.dev/nestjs/bull/blob/f16bf2b31e5c54af73e3f9a2e1619a21ab122970/packages/tsconfig.build.json#L15) file does not have `strict: true`

Sikora00 replied:
if it's optional, can't you just `if` it?

thiagomini replied:
You mean to add an `if` statement before calling `subscribeToEventIfListener` ?

Sikora00 replied:
Yes.
Maybe like 
```ts
if(!wrapper.host) {
  throw new Error()
}
```

thiagomini replied:
Hmmmm I see, but do you think this is really the expected behavior here? To be honest, I'm not sure what would happen if this host really didn't exist when we tried to inject the instance. @kamilmysliwiec  do you have some opinion on that as wel?

kamilmysliwiec replied:
At that time, `host` will never be `undefined` so the `as Module` assertion is correct here.

delucca commented on lib/event-subscribers.loader.ts:
```diff
@@ -31,39 +45,126 @@ export class EventSubscribersLoader
   }
 
-  private subscribeToEventIfListener(
+  private async subscribeToEventIfListener(
```
Why do we need `async` here? I checked the new added code and it seems we don't use `await` anywhere 🤔

thiagomini replied:
Good point, it's not required anymore, I'll fix it!

kamilmysliwiec replied:
Great catch @delucca!

delucca reviewed (commented):
Great job 🚀 

I'm not going to approve it, since I don't know if this could break something. But I added a few minor comments 😄

thiagomini commented on lib/event-subscribers.loader.ts:
```diff
@@ -31,39 +45,126 @@ export class EventSubscribersLoader
+    */
+
+    const payloadObjectOrArray =
+      eventPayload.length > 1 ? eventPayload : eventPayload[0];
```
I'll double check this, but I think the `eventPayload` here will actually always be an array, even if we passed a string. That's because the `eventPayload` here is actually an alias for the function's `args` parameters, which is a rest parameter (and thus, an array)

thiagomini replied:
I added a test for the case you mentioned in [this commit](https://github.com/nestjs/event-emitter/pull/544/commits/afeeb0c134822655c20097ad9daa1d5b132d0a3b) and it works as expected, therefore I think we don't have to check against an array here

kamilmysliwiec commented on lib/event-subscribers.loader.ts:
```diff
@@ -3,20 +3,34 @@ import {
 export class EventSubscribersLoader
   implements OnApplicationBootstrap, OnApplicationShutdown
 {
+  private injector = new Injector();
```
```suggestion
  private readonly injector = new Injector();
```
nit

kamilmysliwiec commented:
Fantastic PR @thiagomini, thank you 🙌

thiagomini commented:
> Fantastic PR @thiagomini, thank you 🙌

Thank you, Kamil! I added a few other commits here to resolve the review suggestions. I also refactored a bit the test code to use `constants` instead of manually set objects. In the final commit, I refactored a bit of the `EventsProviderRequestScopedConsumer` to look cleaner. Before this commit, I was adding more and more static variables to the class to handle different event payloads, so I decided to move this logic away to another object, the `RequestScopedEventPayload` class.

kamilmysliwiec commented on lib/event-subscribers.loader.ts:
```diff
@@ -31,39 +45,126 @@ export class EventSubscribersLoader
+
     const { event, options } = eventListenerMetadata;
-    const listenerMethod = !!options?.prependListener
+    const listenerMethod = this.registerListenerMethodBasedOn(options);
```
```suggestion
    const listenerMethod = this.getRegisterListenerMethodBasedOn(options);
```
